### PR TITLE
[DistillationTrainer refactor] Pin the distillation objective before refactoring

### PR DESCRIPTION
**Pull request by qgallouedec**
*Plan step 0.1 — first PR of the distillation refactor sequence. Part of #6449.*

`DistillationTrainer` is about to go through a long refactor: top-k support removal, then a switch to a chunked loss. The implementation is expected to change; the value it computes is not. This adds the tests that hold that line.

Three pins, all seeded, over `beta` ∈ {0, 0.5, 1}, with and without a label mask:

- **vs. a naive reference implementation** written straight from the definition. It forms probabilities explicitly and builds the mixture in probability space, rather than reusing `F.kl_div`'s inverted argument order and the `logsumexp` trick — so it can actually catch an argument-order or mixture-weight regression rather than reproducing one.
- **vs. `GKDTrainer.generalized_jsd_loss`**, which implements the same objective. This is the cross-trainer contract: if it breaks, either the refactor changed the objective or GKD drifted.
- **the temperature path**, which is scheduled to become sampling-only later in the sequence. Pinning it now makes that a visible diff instead of a silent drift.

No production code touched.

### Note

The existing `TestGeneralizedJSDLoss` fixture builds its logits with an unseeded `torch.randn`, so it cannot pin values — it only checks relations (`loss_beta_0 != loss_beta_1` and similar). Left as is; the new class is seeded and additive.

### Risk

Low — tests only.

---

> [!NOTE]
> **Low Risk**
> Tests-only PR with no runtime or training behavior changes; risk is limited to CI coverage of the distillation objective.
> 
> **Overview**
> Adds **regression tests only** (no production changes) to lock `DistillationTrainer.generalized_jsd_loss` before upcoming refactors (top-k removal, chunked loss).
> 
> Introduces a **naive `_reference_generalized_jsd`** helper (explicit probabilities and mixture in probability space, independent of `F.kl_div` tricks) and a seeded **`TestGeneralizedJSDLossIsPinned`** class that asserts the trainer loss matches that reference for `beta` ∈ {0, 0.25, 1.0}, with and without label masks.
> 
> Also pins a **cross-trainer contract** against `GKDTrainer.generalized_jsd_loss` on the same fixtures, and **explicit temperature=2.0** checks so a future “temperature is sampling-only” change must fail tests visibly. Imports `GKDTrainer` for that comparison.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 911b0e5c7d81da00c9d599d32d9f485117678361. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/experimental/test_distillation_trainer.py [hunk 1]
```diff
@@ -367,6 +368,81 @@ def test_mask_keeps_forward_and_backward_finite(self):
...
+    computes is not. Any diff that moves these numbers is changing the objective and must say so.
+    """
+
+    def setup_method(self):
+        generator = torch.Generator().manual_seed(42)  # seeded: an unseeded fixture cannot pin anything
+        self.student_logits = torch.randn(2, 3, 5, generator=generator)
+        self.teacher_logits = torch.randn(2, 3, 5, generator=generator)
+        self.labels = torch.tensor([[-100, 1, 2], [-100, -100, 3]])
+
+    @pytest.mark.parametrize("beta", [0.0, 0.5, 1.0])
```
Comment (cmpatino) [inline tests/experimental/test_distillation_trainer.py]: I think we should use a value different from 0.5 since that would help us test the asymmetric case of the generalized JSD loss. 

The function is symmetric when `beta=0.5`, so adding a non-symmetric value like 0.25 will help us uncover issues when we swap student and teacher mixture weights.

**Review by cmpatino — COMMENTED** [review]
I included a comment about an additional `beta` value to test the asymmetric case of the JSD loss.

File: tests/experimental/test_distillation_trainer.py (hunk 1 above)
Comment (qgallouedec) [reply tests/experimental/test_distillation_trainer.py]: Thanks, addressed in https://github.com/huggingface/trl/pull/6450/commits/c978997a6f1a2152f3919063166334d4f060bdda
